File: javascript/src/providers/TenMinuteMail.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError, NotFoundError } from '../errors.js';

const BASE = 'https://10minutemail.net';

// Helper to decode Cloudflare obfuscated email
function decodeCfEmail(hex) {
  if (!hex) return '';
  let str = '';
  const k = parseInt(hex.substr(0, 2), 16);
  for (let i = 2; i < hex.length; i += 2) {
    str += String.fromCharCode(parseInt(hex.substr(i, 2), 16) ^ k);
  }
  return str;
}

// Strip HTML tags helper
function stripHtml(html) {
  if (!html) return '';
  return html.replace(/<[^>]+>/g, '').trim();
}

/**
 * 10minutemail.net provider — HTML scraping public mailbox.
 */
export class TenMinuteMail extends TempMailProvider {
  /** @param {{ randomUA?: boolean, useCookies?: boolean }|undefined} config */
  constructor(config = {}) {
    super('10minutemail', { randomUA: config.randomUA, useCookies: true });
    /** @type {string|null} */
    this._email = null;
  }

  /** @returns {Promise<string>} */
  async generateEmail() {
    const res = await this._fetch(`${BASE}/`);
    const html = await res.text();
    const match = html.match(/id="fe_text"[^>]*value="([^"]+)"/i);
    this._email = match ? match[1].trim() : null;
    if (!this._email) {
      throw new TempMailError('No address in response', this.name);
    }
    return this._email;
  }

  /**
   * @param {string} _email
   * @returns {Promise<Message[]>}
   */
  async getInbox(_email) {
    const res = await this._fetch(`${BASE}/mailbox.ajax.php`);
    const html = await res.text();
    const messages = [];
    const rowRe = /<tr[^>]*>([\s\S]*?)<\/tr>/gi;
    
    // Skip header row
    let isHeader = true;
    let rm;
    while ((rm = rowRe.exec(html))) {
      if (isHeader) {
        isHeader = false;
        continue;
      }
      const row = rm[1];
      const cells = [];
      const cellRe = /<td[^>]*>([\s\S]*?)<\/td>/gi;
      let cm;
      while ((cm = cellRe.exec(row))) cells.push(cm[1]);
      if (cells.length < 3) continue;

      // Extract sender (handling Cloudflare obfuscation)
      let sender = '';
      const cfMatch = cells[0].match(/data-cfemail="([^"]+)"/i);
      if (cfMatch) {
        sender = decodeCfEmail(cfMatch[1]);
      } else {
        sender = stripHtml(cells[0]);
      }

      const subject = stripHtml(cells[1]);

      // Extract Date (from title if exists)
      let dateStr = '';
      const dateMatch = cells[2].match(/title="([^"]+)"/i);
      if (dateMatch) {
        dateStr = dateMatch[1];
      } else {
        dateStr = stripHtml(cells[2]);
      }
      const date = new Date(dateStr + (dateStr.toLowerCase().includes('utc') ? '' : ' UTC'));

      // Extract mid (message ID)
      const midMatch = row.match(/mid=([^'&"\s>]+)/i);
      const id = midMatch ? midMatch[1] : '';

      if (id) {
        messages.push(new Message({ id, sender, subject, date }));
      }
    }
    return messages;
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    const mid = messageId.includes(':') ? messageId.split(':')[0] : messageId;
    const res = await this._fetch(`${BASE}/readmail.html?mid=${mid}`);
    const html = await res.text();

    // Check if mail exists (if we get the page)
    const bodyMatch = html.match(/class="mailinhtml"[^>]*>([\s\S]*?)<div[^>]*style="clear:both;"/i);
    if (!bodyMatch) {
      throw new NotFoundError(`Message ${messageId} not found`, this.name);
    }

    const bodyHtmlRaw = bodyMatch[1].trim();

    // Decode Cloudflare email obfuscation in body
    let bodyHtml = bodyHtmlRaw.replace(/<(a|span)[^>]*class="__cf_email__"[^>]*data-cfemail="([^"]+)"[^>]*>([\s\S]*?)<\/\1>/gi, (m, tag, hex) => {
      return decodeCfEmail(hex);
    });
    bodyHtml = bodyHtml.replace(/href="\/cdn-cgi\/l\/email-protection#([^"]+)"/g, (m, hex) => {
      return `href="mailto:${decodeCfEmail(hex)}"`;
    });

    const bodyText = stripHtml(bodyHtml);

    const subjectMatch = html.match(/<div class="mail_header">[\s\S]*?<h2[^>]*>([\s\S]*?)<\/h2>/i);
    const subject = subjectMatch ? stripHtml(subjectMatch[1]) : '';

    const fromMatch = html.match(/<span class="mail_from">([\s\S]*?)<\/span>/i);
    let sender = '';
    if (fromMatch) {
      const cfFrom = fromMatch[1].match(/data-cfemail="([^"]+)"/i);
      sender = cfFrom ? decodeCfEmail(cfFrom[1]) : stripHtml(fromMatch[1]);
    }

    return new MessageDetail({
      id: messageId,
      sender,
      subject,
      date: new Date(), // fallback
      bodyText,
      bodyHtml,
      attachments: [],
    });
  }

  /**
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    this._email = null;
    return true;
  }
}

File: go/providers/ten_minute_mail.go
package providers

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

const tenMinuteMailBaseURL = "https://10minutemail.net"

// TenMinuteMail is the 10minutemail.net provider (HTML scraping public mailbox).
type TenMinuteMail struct {
	client *HttpClient
	email  string
}

// NewTenMinuteMail creates a new TenMinuteMail provider.
func NewTenMinuteMail(config map[string]string) *TenMinuteMail {
	proxies := ParseProxies(config["proxies"])
	randomUA := config["randomUa"] == "true"
	return &TenMinuteMail{client: NewHttpClient(proxies, randomUA, true)}
}

// Decode Cloudflare obfuscated email address
func decodeCfEmail(hex string) string {
	if len(hex) < 4 {
		return ""
	}
	k, err := strconv.ParseInt(hex[0:2], 16, 64)
	if err != nil {
		return ""
	}
	var builder strings.Builder
	for i := 2; i < len(hex); i += 2 {
		val, err := strconv.ParseInt(hex[i:i+2], 16, 64)
		if err != nil {
			continue
		}
		builder.WriteByte(byte(val ^ k))
	}
	return builder.String()
}

func (t *TenMinuteMail) GenerateEmail() (string, error) {
	body, err := t.client.Get(tenMinuteMailBaseURL + "/")
	if err != nil {
		return "", &TempMailError{Provider: "10minutemail", Message: err.Error()}
	}
	html := string(body)
	re := regexp.MustCompile(`(?i)id="fe_text"[^>]*value="([^"]+)"`)
	match := re.FindStringSubmatch(html)
	if len(match) < 2 {
		return "", &TempMailError{Provider: "10minutemail", Message: "no address in response"}
	}
	t.email = strings.TrimSpace(match[1])
	return t.email, nil
}

func (t *TenMinuteMail) GetInbox(email string) ([]Message, error) {
	body, err := t.client.Get(tenMinuteMailBaseURL + "/mailbox.ajax.php")
	if err != nil {
		return nil, &TempMailError{Provider: "10minutemail", Message: err.Error()}
	}
	html := string(body)
	
	rowRe := regexp.MustCompile(`(?is)<tr[^>]*>(.*?)</tr>`)
	cellRe := regexp.MustCompile(`(?is)<td[^>]*>(.*?)</td>`)
	cfRe := regexp.MustCompile(`(?i)data-cfemail="([^"]+)"`)
	titleRe := regexp.MustCompile(`(?i)title="([^"]+)"`)
	midRe := regexp.MustCompile(`(?i)mid=([^'&"\s>]+)`)

	rows := rowRe.FindAllStringSubmatch(html, -1)
	if len(rows) <= 1 {
		return []Message{}, nil
	}

	var msgs []Message
	// Skip header row
	for i := 1; i < len(rows); i++ {
		rowHtml := rows[i][1]
		cells := cellRe.FindAllStringSubmatch(rowHtml, -1)
		if len(cells) < 3 {
			continue
		}

		// Sender
		sender := ""
		cfMatch := cfRe.FindStringSubmatch(cells[0][1])
		if len(cfMatch) >= 2 {
			sender = decodeCfEmail(cfMatch[1])
		} else {
			sender = StripHTML(cells[0][1])
		}

		subject := StripHTML(cells[1][1])

		// Date
		dateStr := ""
		dateMatch := titleRe.FindStringSubmatch(cells[2][1])
		if len(dateMatch) >= 2 {
			dateStr = dateMatch[1]
		} else {
			dateStr = StripHTML(cells[2][1])
		}
		
		// Parse date (usually in UTC)
		if !strings.Contains(strings.ToLower(dateStr), "utc") {
			dateStr += " UTC"
		}
		parsedDate, err := time.Parse("2006-01-02 15:04:05 MST", dateStr)
		if err != nil {
			parsedDate = time.Now()
		}

		// Message ID
		midMatch := midRe.FindStringSubmatch(rowHtml)
		if len(midMatch) < 2 {
			continue
		}
		id := midMatch[1]

		msgs = append(msgs, Message{
			ID:      id,
			Sender:  sender,
			Subject: subject,
			Date:    parsedDate,
		})
	}

	return msgs, nil
}

func (t *TenMinuteMail) ReadMessage(messageID string) (*MessageDetail, error) {
	// composite messageID handling
	mid := messageID
	if idx := strings.Index(messageID, ":"); idx >= 0 {
		mid = messageID[:idx]
	}

	body, err := t.client.Get(tenMinuteMailBaseURL + "/readmail.html?mid=" + mid)
	if err != nil {
		return nil, &TempMailError{Provider: "10minutemail", Message: err.Error()}
	}
	html := string(body)

	bodyRe := regexp.MustCompile(`(?is)class="mailinhtml"[^>]*>(.*?)<div[^>]*style="clear:both;"`)
	bodyMatch := bodyRe.FindStringSubmatch(html)
	if len(bodyMatch) < 2 {
		return nil, &NotFoundError{Provider: "10minutemail", Resource: "message " + messageID}
	}

	bodyHtml := strings.TrimSpace(bodyMatch[1])

	// Decode Cloudflare obfuscated email links inside body
	cfLinkRe := regexp.MustCompile(`(?i)<(?:a|span)[^>]*class="__cf_email__"[^>]*data-cfemail="([^"]+)"[^>]*>.*?<\/(?:a|span)>`)
	bodyHtml = cfLinkRe.ReplaceAllStringFunc(bodyHtml, func(m string) string {
		cfMatch := cfLinkRe.FindStringSubmatch(m)
		if len(cfMatch) >= 2 {
			return decodeCfEmail(cfMatch[1])
		}
		return m
	})

	cfProtectionRe := regexp.MustCompile(`(?i)href="/cdn-cgi/l/email-protection#([^"]+)"`)
	bodyHtml = cfProtectionRe.ReplaceAllStringFunc(bodyHtml, func(m string) string {
		cfMatch := cfProtectionRe.FindStringSubmatch(m)
		if len(cfMatch) >= 2 {
			return `href="mailto:` + decodeCfEmail(cfMatch[1]) + `"`
		}
		return m
	})

	bodyText := StripHTML(bodyHtml)

	// Extract Subject
	subjectRe := regexp.MustCompile(`(?is)<div class="mail_header">.*?<h2[^>]*>(.*?)</h2>`)
	subjectMatch := subjectRe.FindStringSubmatch(html)
	subject := ""
	if len(subjectMatch) >= 2 {
		subject = StripHTML(subjectMatch[1])
	}

	// Extract Sender
	fromRe := regexp.MustCompile(`(?is)<span class="mail_from">(.*?)</span>`)
	fromMatch := fromRe.FindStringSubmatch(html)
	sender := ""
	if len(fromMatch) >= 2 {
		cfRe := regexp.MustCompile(`(?i)data-cfemail="([^"]+)"`)
		cfFrom := cfRe.FindStringSubmatch(fromMatch[1])
		if len(cfFrom) >= 2 {
			sender = decodeCfEmail(cfFrom[1])
		} else {
			sender = StripHTML(fromMatch[1])
		}
	}

	return &MessageDetail{
		Message: Message{
			ID:      mid,
			Sender:  sender,
			Subject: subject,
			Date:    time.Now(),
		},
		BodyText: bodyText,
		BodyHTML: bodyHtml,
	}, nil
}

func (t *TenMinuteMail) DeleteEmail(email string) (bool, error) {
	t.email = ""
	return true, nil
}

func (t *TenMinuteMail) WaitForEmail(email string, timeout time.Duration, interval time.Duration) (*Message, error) {
	return waitForEmail(t, email, timeout, interval)
}
